### PR TITLE
68 Support Query Operators for Entities

### PR DESCRIPTION
**Pull request by subhashb**
Supported filtering/query operators are:

- `exact` - Default operator, equivalent to eq for all objects
- `iexact` - case-insensitive match, if String. Will throw error for non-character types
- `contains`
- `icontains`
- `in` - Applicable to iterable types
- `gte`
- `gt`
- `lte`
- `lt`

**Notes:**

- Negation is supported by `exclude()`: Criteria are ANDed and then negated
- All criteria specified in `.filter()` are ANDed
- Multiple filter clauses get collapsed together as part of a single AND clause

**Examples:**
* _Example 1_: `Account.query.filter(first_name__iexact='Subhash', last_name__iexact=='Bhushan')
* _Example 2_: `Account.query.filter(first_name__iexact='Subhash').filter(last_name__iexact=='Bhushan')
* _Example 3_: `Account.query.filter(first_name__iexact='Subhash', last_name__iexact=='Bhushan').exclude(full_name__iexact='Subhash Bhushan)
* _Example 4_: `Account.query.filter(first_name__iexact='Subhash', last_name__contains='Bhushan').order_by('email')`

The Sample Dict repository implementation has been updated to support these operations.

### CODE REVIEW

File: src/protean/impl/repository/dict_repo.py [hunk 1]
```diff
@@ -60,21 +61,15 @@ def _filter_objects(self, page: int = 1, per_page: int = 10,
...
             for fk, fv in filters.items():
-                if type(fv) in (list, tuple):
-                    if item[fk] not in fv:
-                        match = False
-                else:
-                    if item[fk] != fv:
-                        match = False
+                lookup_class = self.get_lookup(fv[0])
+                lookup = lookup_class(item[fk], fv[1])
+                match &= eval(lookup.as_expression())
```
Comment (advishnuprasad) [inline src/protean/impl/repository/dict_repo.py]: Do we have a way other than `eval`? 
btw, what is `&=` ?

Comment (subhashb) [reply src/protean/impl/repository/dict_repo.py]: `&` is bitwise operator, and `&=` does the operation and assigns the value back (https://www.tutorialspoint.com/python/bitwise_operators_example.htm)

```
Python 3.6.5 (default, Apr 28 2018, 22:32:18)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> match = True
>>> match &= True
>>> print(match)
True
>>> match &= False
>>> print(match)
False
>>> match = False
>>> match &= True
>>> print(match)
False
>>> match &= False
>>> print(match)
False
```

Comment (subhashb) [reply src/protean/impl/repository/dict_repo.py]: In short, `match` remains `True` only if all future comparisons are True. If any of them are False, `match` will become `False` and remain `False`.

File: src/protean/core/repository/base.py [hunk 2]
```diff
@@ -112,3 +113,25 @@ def get_connection(self):
     @abstractmethod
     def close_connection(self, conn):
         """ Close the connection object for the repository"""
+
+
+class Lookup(metaclass=ABCMeta):
+    """Base Lookup class to implement in Adapters"""
+    lookup_name = None
+
+    def __init__(self, source, target):
```
Comment (abhishek-ram) [inline src/protean/core/repository/base.py]: Can we add the adapter class to the lookup initialization.

File: src/protean/core/entity.py [hunk 3]
```diff
@@ -138,6 +138,8 @@ def __init__(self, entity_cls_name: str, page: int = 1, per_page: int = 10,
         self._filters = filters
 
         self._result_cache = None
+        self._restructured_filters = {}
```
Comment (abhishek-ram) [inline src/protean/core/entity.py]: We don't really need to save this in the object, if we ever need it we can just call the function to get it back

File: src/protean/core/entity.py [hunk 4]
```diff
@@ -214,9 +232,14 @@ def all(self):
         # order_by clause must be list of keys
         order_by = model_cls.opts_.order_by if not self._order_by else self._order_by
 
+        # Breakdown filters
+        self._restructured_filters = self._restructure_filters(adapter, self._filters)
```
Comment (abhishek-ram) [inline src/protean/core/entity.py]: We don't really need to save this in the object, if we ever need it we can just call the function to get it back. We can then just call the attrs as `filters` and `excludes`

**Pull request merged**